### ISSUE TITLE
open2 oddity with eval "..." and STDIN closed

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#8088]
Migrated from [rt.perl.org#37033](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=37033) (status was 'resolved')

Searchable as RT37033$

**Comment by p5pRT** [comment]
### From pajas@matfyz.cz

This is a bug report for perl from pajas@&#8203;matfyz\.cz\,
generated with the help of perlbug 1\.35 running under perl v5\.8\.6\.




---
Hi\!

A virtual beer to anyone who sheds some light on this&#8203;:\-\)

I'm puzzled by a weird behavior of output redirection in a larger
application of mine\. Finally\, I was able to trim the code down to the
few lines below\.

\_\_CODE\_STARTS\_HERE\_\_
\#\!/usr/bin/env perl
close STDIN;
eval \<\<'EVAL'; \# \<\- remove or replace with eval \{\}
&nbsp;               \#    to see the correct output
&nbsp;  use Encode&#8203;::Byte; \# \<\- instead of POD of this
&nbsp;  use IPC&#8203;::Open2;
&nbsp;  open2\('>&STDOUT'\,$P\,'cat'\) || die "cannot open pipe&#8203;: $\!\\n";
&nbsp;  print $P \("HALLO\\n"\);
EVAL
\_\_END\_\_

It is now only a way round to pass the string "HALLO\\n" to 'cat'
program\. However\, under the specific scenario\, the content that follows
\_\_END\_\_ in the module Encode/Byte\.pm is unexpectedly passed to 'cat'
instead\.

HELLO is printed if either of the following conditions is satisfied&#8203;:

 \*  eval "" is removed 
 \*  eval "" is replaced with eval \{ \}
 \*  the initial close STDIN is removed
 \*  close STDIN is moved into the eval

If 'use Encode&#8203;::Byte' is removed\, the program outputs nothing\.

I'm assuming this is a bug because I don't see a reason why changing
eval"" to eval\{\} should make any difference here and also why closing
STDIN should be relevant\.

I'm marking the bug as medium severe because of the fact that 'use
Module' exposes an open filehandle \(fd0\) with an unexpected content and open2
passes it to a subprocess where it shouldn't\.

Beside the reported versions\, I was able to reproduce the problem with
a 5\.9\.3 build on Fedora Core 3\.

Thanks\,

\-\- Petr



---


---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=medium


---
This perlbug was built using Perl v5\.8\.6 \- Sat Mar 19 17&#8203;:33&#8203;:54 UTC 2005
It is being executed now by  Perl v5\.8\.6 \- Sat Mar 19 17&#8203;:29&#8203;:44 UTC 2005\.

Site configuration information for perl v5\.8\.6&#8203;:

Configured by abuild at Sat Mar 19 17&#8203;:29&#8203;:44 UTC 2005\.

Summary of my perl5 \(revision 5 version 8 subversion 6\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=linux\, osvers=2\.6\.9\, archname=i586\-linux\-thread\-multi
&nbsp;    uname='linux salieri 2\.6\.9 \#1 smp fri jan 14 15&#8203;:41&#8203;:33 utc 2005 i686 athlon 
i386 gnulinux '
&nbsp;    config\_args='\-ds \-e \-Dprefix=/usr \-Dvendorprefix=/usr \-Dinstallusrbinperl 
\-Dusethreads \-Di\_db \-Di\_dbm \-Di\_ndbm \-Di\_gdbm \-Duseshrplib=true 
\-Doptimize=\-O2 \-march=i586 \-mcpu=i686 \-fmessage\-length=0 \-Wall \-g \-Wall 
\-pipe'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=define use5005threads=undef useithreads=define 
usemultiplicity=define
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS 
\-DDEBUGGING \-fno\-strict\-aliasing \-pipe \-D\_LARGEFILE\_SOURCE 
\-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O2 \-march=i586 \-mcpu=i686 \-fmessage\-length=0 \-Wall \-g \-Wall 
\-pipe'\,
&nbsp;    cppflags='\-D\_REENTRANT \-D\_GNU\_SOURCE \-DTHREADS\_HAVE\_PIDS \-DDEBUGGING 
\-fno\-strict\-aliasing \-pipe'
&nbsp;    ccversion=''\, gccversion='3\.3\.5 20050117 \(prerelease\) \(SUSE Linux\)'\, 
gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=12
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, 
lseeksize=8
&nbsp;    alignbytes=4\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='cc'\, ldflags =''
&nbsp;    libpth=/lib /usr/lib /usr/local/lib
&nbsp;    libs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    perllibs=\-lnsl \-ldl \-lm \-lcrypt \-lutil \-lpthread \-lc
&nbsp;    libc=\, so=so\, useshrplib=true\, libperl=libperl\.so
&nbsp;    gnulibc\_version='2\.3\.4'
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags='\-Wl\,\-E 
\-Wl\,\-rpath\,/usr/lib/perl5/5\.8\.6/i586\-linux\-thread\-multi/CORE'
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-shared'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl v5\.8\.6&#8203;:
&nbsp;    /usr/lib/perl5/5\.8\.6/i586\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/5\.8\.6
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.6/i586\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/site\_perl/5\.8\.6
&nbsp;    /usr/lib/perl5/site\_perl
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.6/i586\-linux\-thread\-multi
&nbsp;    /usr/lib/perl5/vendor\_perl/5\.8\.6
&nbsp;    /usr/lib/perl5/vendor\_perl
&nbsp;    \.



---
Environment for perl v5\.8\.6&#8203;:
&nbsp;    HOME=/home/pajas
&nbsp;    LANG=cs\_CZ\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    
PATH=/home/pajas/bin&#8203;:/usr/local/bin&#8203;:/usr/bin&#8203;:/usr/X11R6/bin&#8203;:/bin&#8203;:/usr/games&#8203;:/opt/gnome/bin&#8203;:/opt/kde3/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @jkeenan

On Tue Aug 30 12&#8203;:04&#8203;:45 2005\, pajas wrote&#8203;:
> This is a bug report for perl from pajas@&#8203;matfyz\.cz\,
> generated with the help of perlbug 1\.35 running under perl v5\.8\.6\.
> 

> A virtual beer to anyone who sheds some light on this&#8203;:\-\)
> 
> I'm puzzled by a weird behavior of output redirection in a larger
> application of mine\. Finally\, I was able to trim the code down to the
> few lines below\.
> 
> \_\_CODE\_STARTS\_HERE\_\_
> \#\!/usr/bin/env perl
> close STDIN;
> eval \<\<'EVAL'; \# \<\- remove or replace with eval \{\}
>                \#    to see the correct output
>   use Encode&#8203;::Byte; \# \<\- instead of POD of this
>   use IPC&#8203;::Open2;
>   open2\('>&STDOUT'\,$P\,'cat'\) || die "cannot open pipe&#8203;: $\!\\n";
>   print $P \("HALLO\\n"\);
> EVAL
> \_\_END\_\_
> 
> It is now only a way round to pass the string "HALLO\\n" to 'cat'
> program\. However\, under the specific scenario\, the content that
> follows
> \_\_END\_\_ in the module Encode/Byte\.pm is unexpectedly passed to 'cat'
> instead\.
> 
> HELLO is printed if either of the following conditions is satisfied&#8203;:
> 
>  \*  eval "" is removed
>  \*  eval "" is replaced with eval \{ \}
>  \*  the initial close STDIN is removed
>  \*  close STDIN is moved into the eval
> 
> If 'use Encode&#8203;::Byte' is removed\, the program outputs nothing\.
> 
> I'm assuming this is a bug because I don't see a reason why changing
> eval"" to eval\{\} should make any difference here and also why closing
> STDIN should be relevant\.
> 
> I'm marking the bug as medium severe because of the fact that 'use
> Module' exposes an open filehandle \(fd0\) with an unexpected content
> and open2
> passes it to a subprocess where it shouldn't\.
> 
> Beside the reported versions\, I was able to reproduce the problem with
> a 5\.9\.3 build on Fedora Core 3\.
> 

And I can reproduce the different variants with Perl 5\.14\.2 on Darwin\.

jimk

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @iabyn

On Tue\, Jan 03\, 2012 at 05&#8203;:58&#8203;:33PM \-0800\, James E Keenan via RT wrote&#8203;:
> On Tue Aug 30 12&#8203;:04&#8203;:45 2005\, pajas wrote&#8203;:
> > This is a bug report for perl from pajas@&#8203;matfyz\.cz\,
> > generated with the help of perlbug 1\.35 running under perl v5\.8\.6\.
> > 
> 
> > A virtual beer to anyone who sheds some light on this&#8203;:\-\)
> > 
> > I'm puzzled by a weird behavior of output redirection in a larger
> > application of mine\. Finally\, I was able to trim the code down to the
> > few lines below\.
> > 
> > \_\_CODE\_STARTS\_HERE\_\_
> > \#\!/usr/bin/env perl
> > close STDIN;
> > eval \<\<'EVAL'; \# \<\- remove or replace with eval \{\}
> >                \#    to see the correct output
> >   use Encode&#8203;::Byte; \# \<\- instead of POD of this
> >   use IPC&#8203;::Open2;
> >   open2\('>&STDOUT'\,$P\,'cat'\) || die "cannot open pipe&#8203;: $\!\\n";
> >   print $P \("HALLO\\n"\);
> > EVAL
> > \_\_END\_\_
> > 
> > It is now only a way round to pass the string "HALLO\\n" to 'cat'
> > program\. However\, under the specific scenario\, the content that
> > follows
> > \_\_END\_\_ in the module Encode/Byte\.pm is unexpectedly passed to 'cat'
> > instead\.
> > 
> > HELLO is printed if either of the following conditions is satisfied&#8203;:
> > 
> >  \*  eval "" is removed
> >  \*  eval "" is replaced with eval \{ \}
> >  \*  the initial close STDIN is removed
> >  \*  close STDIN is moved into the eval
> > 
> > If 'use Encode&#8203;::Byte' is removed\, the program outputs nothing\.
> > 
> > I'm assuming this is a bug because I don't see a reason why changing
> > eval"" to eval\{\} should make any difference here and also why closing
> > STDIN should be relevant\.
> > 
> > I'm marking the bug as medium severe because of the fact that 'use
> > Module' exposes an open filehandle \(fd0\) with an unexpected content
> > and open2
> > passes it to a subprocess where it shouldn't\.
> > 
> > Beside the reported versions\, I was able to reproduce the problem with
> > a 5\.9\.3 build on Fedora Core 3\.
> > 
> 
> And I can reproduce the different variants with Perl 5\.14\.2 on Darwin\.

It's not a bug\.
The eval is only significant in that it allows a 'use' to happen \*after\*
STDIN has been closed\.

A similar effect can be shown here&#8203;:

&nbsp;    BEGIN \{ close STDIN; \}
&nbsp;    use strict;
&nbsp;    while \(\<>\) \{
&nbsp;    chomp;
&nbsp;    print "\[$\_\]\\n";
&nbsp;    \}

which prints out all the lines in lib/strict\.pm following the \_\_END\_\_\.

When perl opens a source file\, it uses the next available file descriptor
\-\- which happens to be 0 in this case\, since STDIN was closed\. Since the
source file contains an \_\_END\_\_\, perl keeps the source file open\, in case
of a later \<strict&#8203;::DATA>\. So reading from STDIN \(aka fd 0\) reads from the
text following \_\_END\_\_\.

\-\- 
Diplomacy is telling someone to go to hell in such a way that they'll
look forward to the trip

**Comment by p5pRT** [comment]
### From @jkeenan

On 1/8/12 9&#8203;:56 AM\, Dave Mitchell wrote&#8203;:
>
> It's not a bug\.
> The eval is only significant in that it allows a 'use' to happen \*after\*
> STDIN has been closed\.
>
> A similar effect can be shown here&#8203;:
>
>      BEGIN \{ close STDIN; \}
>      use strict;
>      while \(\<>\) \{
>     chomp;
>     print "\[$\_\]\\n";
>      \}
>
> which prints out all the lines in lib/strict\.pm following the \_\_END\_\_\.
>
> When perl opens a source file\, it uses the next available file descriptor
> \-\- which happens to be 0 in this case\, since STDIN was closed\. Since the
> source file contains an \_\_END\_\_\, perl keeps the source file open\, in case
> of a later\<strict&#8203;::DATA>\. So reading from STDIN \(aka fd 0\) reads from the
> text following \_\_END\_\_\.
>

Is there anything which should be added to the Perl documentation with 
respect to this?

Otherwise\, we should close the ticket\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @rjbs

I think this falls under the category of "not to be documented\."  It's standard fd behavior\, and 
while it can be surprising\, we can either document it in one place that won't be helpful or we can  
put documentation all over the place that will just confuse readers\.

I think this is a pretty rare scenario that is actually going to be best helped by the user asking 
what's up\.

tl;dr\, some edges are best left explained anew each time

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

I think this falls under the category of "not to be documented\."  It's standard fd behavior\, and 
while it can be surprising\, we can either document it in one place that won't be helpful or we can  
put documentation all over the place that will just confuse readers\.

I think this is a pretty rare scenario that is actually going to be best helped by the user asking 
what's up\.

tl;dr\, some edges are best left explained anew each time

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'rejected'

**Comment by p5pRT** [comment]
### From @nwc10

On Sun\, Jan 08\, 2012 at 02&#8203;:56&#8203;:12PM \+0000\, Dave Mitchell wrote&#8203;:

> It's not a bug\.

I think it \*is\* a bug :\-\)

> The eval is only significant in that it allows a 'use' to happen \*after\*
> STDIN has been closed\.
> 
> A similar effect can be shown here&#8203;:
> 
>     BEGIN \{ close STDIN; \}
>     use strict;
>     while \(\<>\) \{
>     chomp;
>     print "\[$\_\]\\n";
>     \}
> 
> which prints out all the lines in lib/strict\.pm following the \_\_END\_\_\.
> 
> When perl opens a source file\, it uses the next available file descriptor
> \-\- which happens to be 0 in this case\, since STDIN was closed\. Since the
> source file contains an \_\_END\_\_\, perl keeps the source file open\, in case
> of a later \<strict&#8203;::DATA>\. So reading from STDIN \(aka fd 0\) reads from the
> text following \_\_END\_\_\.

perldata says&#8203;:

&nbsp;    Text after \_\_DATA\_\_ may be read via the filehandle C\<PACKNAME&#8203;::DATA>\,
&nbsp;    where C\<PACKNAME> is the package that was current when the \_\_DATA\_\_
&nbsp;    token was encountered\.  The filehandle is left open pointing to the
&nbsp;    line after \_\_DATA\_\_\.  It is the program's responsibility to
&nbsp;    C\<close DATA> when it is done reading from it\.  For compatibility with
&nbsp;    older scripts written before \_\_DATA\_\_ was introduced\, \_\_END\_\_ behaves
&nbsp;    like \_\_DATA\_\_ in the top level script \(but not in files loaded with
&nbsp;    C\<require> or C\<do>\) and leaves the remaining contents of the
&nbsp;    file accessible via C\<main&#8203;::DATA>\.


ie the behavour of leaving the file handle open would be correct for \_\_DATA\_\_
But this is \_\_END\_\_\, and it's not the top level script\, so it should be
closed\.

Contrast what happens when one causes file descriptor 1 to be the one
used to read lib/strict\.pm&#8203;:

$ strace \-eopen\,close \./perl \-Ilib \-e 'close STDOUT; eval "use strict"; '
open\("/etc/ld\.so\.cache"\, O\_RDONLY\)      = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libnsl\.so\.1"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libdl\.so\.2"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libm\.so\.6"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libcrypt\.so\.1"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libutil\.so\.1"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libpthread\.so\.0"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libc\.so\.6"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/dev/urandom"\, O\_RDONLY|O\_LARGEFILE\) = 3
close\(3\)                                = 0
open\("/dev/null"\, O\_RDONLY|O\_LARGEFILE\) = 3
close\(3\)                                = 0
close\(1\)                                = 0
open\("lib/strict\.pm"\, O\_RDONLY|O\_LARGEFILE\) = 1
close\(1\)                                = 0

with file descriptor 0&#8203;:

$ strace \-eopen\,close \./perl \-Ilib \-e 'close STDIN; eval "use strict"; '
open\("/etc/ld\.so\.cache"\, O\_RDONLY\)      = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libnsl\.so\.1"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libdl\.so\.2"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libm\.so\.6"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libcrypt\.so\.1"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libutil\.so\.1"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libpthread\.so\.0"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/lib/i686/cmov/libc\.so\.6"\, O\_RDONLY\) = 3
close\(3\)                                = 0
open\("/dev/urandom"\, O\_RDONLY|O\_LARGEFILE\) = 3
close\(3\)                                = 0
open\("/dev/null"\, O\_RDONLY|O\_LARGEFILE\) = 3
close\(3\)                                = 0
close\(0\)                                = 0
open\("lib/strict\.pm"\, O\_RDONLY|O\_LARGEFILE\) = 0


There's no final close\(0\) there\.

I believe it's down to this piece of special case code in
Perl\_lex\_next\_chunk\(\)&#8203;:

&nbsp;    /\* End of real input\.  Close filehandle \(unless it was STDIN\)\,
&nbsp;     \* then add implicit termination\.
&nbsp;     \*/
&nbsp;    if \(\(PerlIO\*\)PL\_parser\->rsfp == PerlIO\_stdin\(\)\)
&nbsp;        PerlIO\_clearerr\(PL\_parser\->rsfp\);
&nbsp;    else if \(PL\_parser\->rsfp\)
&nbsp;        \(void\)PerlIO\_close\(PL\_parser\->rsfp\);

which I guess should \*ideally\* be deciding to close the file handle based
on whether it was opened for the parse\, not on whether it happens to be
re\-using the stdin slot\. Which would mean storing state as to whether
PL\_parser\->rsfp was "owned" by the parser struct \(ie opened by the parser\)\,
or something passed in by an external owner\.

The comment dates from commit f0e67a1d29102aa9\. The previous version of the
comment was&#8203;:

&nbsp;          /\* Close the filehandle\.  Could be from
&nbsp;           \* STDIN\, or a regular file\.  If we were reading code from
&nbsp;           \* STDIN \(because the commandline held no \-e or filename\)
&nbsp;           \* then we don't close it\, we reset it so the code can
&nbsp;           \* read from STDIN too\.
&nbsp;           \*/


which \(I think\) agrees with my assessment \- the intent of the check as
given by the comment isn't the same as what the code actually tests\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @iabyn

On Mon\, Jan 09\, 2012 at 01&#8203;:36&#8203;:37PM \+0000\, Nicholas Clark wrote&#8203;:
> On Sun\, Jan 08\, 2012 at 02&#8203;:56&#8203;:12PM \+0000\, Dave Mitchell wrote&#8203;:
> 
> > It's not a bug\.
> 
> I think it \*is\* a bug :\-\)

Ah\, so it is :\-\(\. I'll mark the ticket as open again


\-\- 
Wesley Crusher gets beaten up by his classmates for being a smarmy git\,
and consequently has a go at making some friends of his own age for a
change\.
&nbsp;    \-\- Things That Never Happen in "Star Trek" \#18

**Comment by p5pRT** [comment]
@iabyn - Status changed from 'rejected' to 'open'

**Comment by p5pRT** [comment]
### From @nwc10

On Fri Jan 13 06&#8203;:03&#8203;:24 2012\, davem wrote&#8203;:
> On Mon\, Jan 09\, 2012 at 01&#8203;:36&#8203;:37PM \+0000\, Nicholas Clark wrote&#8203;:
> > On Sun\, Jan 08\, 2012 at 02&#8203;:56&#8203;:12PM \+0000\, Dave Mitchell wrote&#8203;:
> > 
> > > It's not a bug\.
> > 
> > I think it \*is\* a bug :\-\)
> 
> Ah\, so it is :\-\(\. I'll mark the ticket as open again

I've pushed a proposed fix to the branch smoke\-me/RT37033

Nicholas Clark

**Comment by p5pRT** [comment]
### From @xdg

On Sun\, Jan 8\, 2012 at 7&#8203;:09 PM\, Ricardo SIGNES via RT
\<perlbug\-comment@&#8203;perl\.org> wrote&#8203;:
>
> tl;dr\, some edges are best left explained anew each time

I think the existing documentation could be stronger without getting
into the details\, however\.

perldata now&#8203;:

&nbsp;   \.\.\.The filehandle is left open pointing to the
&nbsp;   line after \_\_DATA\_\_\.  It is the program's responsibility to
&nbsp;   C\<close DATA> when it is done reading from it\.

Perhaps that could be&#8203;:

&nbsp;   \.\.\.The filehandle is left open pointing to the
&nbsp;   line after \_\_DATA\_\_\.  The program should
&nbsp;   C\<close DATA> when it is done reading from it\.

Leaving it open leaks filehandles if the module is reloaded for any
reason\, so it's a safer practice to close it\.

I'm not sure if we need to explain why since reloading is pretty rare\,
but we could with a single sentence like that if we chose to\.

\-\- David

**Comment by p5pRT** [comment]
### From @nwc10

On Fri Feb 24 07&#8203;:37&#8203;:19 2012\, nicholas wrote&#8203;:
> On Fri Jan 13 06&#8203;:03&#8203;:24 2012\, davem wrote&#8203;:
> > On Mon\, Jan 09\, 2012 at 01&#8203;:36&#8203;:37PM \+0000\, Nicholas Clark wrote&#8203;:
> > > On Sun\, Jan 08\, 2012 at 02&#8203;:56&#8203;:12PM \+0000\, Dave Mitchell wrote&#8203;:
> > > 
> > > > It's not a bug\.
> > > 
> > > I think it \*is\* a bug :\-\)
> > 
> > Ah\, so it is :\-\(\. I'll mark the ticket as open again
> 
> I've pushed a proposed fix to the branch smoke\-me/RT37033

Merged to blead as a907031882152818

Pondering David's related suggested documentation tweak\, in the comments
on this ticket\. I think we should resolve that or move it to a new
ticket before marking this ticket resolved\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Feb 27 02&#8203;:49&#8203;:24 2012\, nicholas wrote&#8203;:
> 
> Pondering David's related suggested documentation tweak\, in the comments
> on this ticket\. I think we should resolve that or move it to a new
> ticket before marking this ticket resolved\.
> 
> Nicholas Clark

In the interest of getting this ticket closed without opening a separate
ticket for a documentation patch\, I have worked David Golden's
suggestion into the patch attached\.

Thank you very much\.
Jim Keenan

**Comment by p5pRT** [comment]
### From @jkeenan

<details><summary>0001-Explain-why-program-must-close-__DATA__-handle.patch</summary>

```diff
From 901a54241c6ff1423fc54301abddbf361c9139c3 Mon Sep 17 00:00:00 2001
From: jkeenan <jkeenan@cpan.org>
Date: Mon, 27 Feb 2012 12:34:10 -0500
Subject: [PATCH] Explain why program must close __DATA__ handle.

Adapted from suggestion by David Golden++.  For RT #37033.
---
 pod/perldata.pod |   18 +++++++++---------
 1 files changed, 9 insertions(+), 9 deletions(-)

diff --git a/pod/perldata.pod b/pod/perldata.pod
index 16ceb41..d14ac10 100644
--- a/pod/perldata.pod
+++ b/pod/perldata.pod
@@ -413,15 +413,15 @@ The two control characters ^D and ^Z, and the tokens __END__ and __DATA__
 may be used to indicate the logical end of the script before the actual
 end of file.  Any following text is ignored.
 
-Text after __DATA__ may be read via the filehandle C<PACKNAME::DATA>,
-where C<PACKNAME> is the package that was current when the __DATA__
-token was encountered.  The filehandle is left open pointing to the
-line after __DATA__.  It is the program's responsibility to
-C<close DATA> when it is done reading from it.  For compatibility with
-older scripts written before __DATA__ was introduced, __END__ behaves
-like __DATA__ in the top level script (but not in files loaded with
-C<require> or C<do>) and leaves the remaining contents of the
-file accessible via C<main::DATA>.
+Text after __DATA__ may be read via the filehandle C<PACKNAME::DATA>, where
+C<PACKNAME> is the package that was current when the __DATA__ token was
+encountered.  The filehandle is left open pointing to the line after __DATA__.
+The program should C<close DATA> when it is done reading from it. (Leaving it
+open leaks filehandles if the module is reloaded for any reason, so it's a
+safer practice to close it.) For compatibility with older scripts written
+before __DATA__ was introduced, __END__ behaves like __DATA__ in the top level
+script (but not in files loaded with C<require> or C<do>) and leaves the
+remaining contents of the file accessible via C<main::DATA>.
 
 See L<SelfLoader> for more description of __DATA__, and
 an example of its use.  Note that you cannot read from the DATA
-- 
1.6.3.2


```
</details>

**Comment by p5pRT** [comment]
### From @rjbs

Thanks\, applied as a6df44c751fa4d86cddb76cc0a040779fe2e7319 \(unless revised for porting 
tests before pushing upstream\)\.

**Comment by p5pRT** [comment]
### From [Unknown Contact. See original ticket]

Thanks\, applied as a6df44c751fa4d86cddb76cc0a040779fe2e7319 \(unless revised for porting 
tests before pushing upstream\)\.

**Comment by p5pRT** [comment]
@rjbs - Status changed from 'open' to 'resolved'
